### PR TITLE
Copy file with correct permissions and owner

### PR DESCRIPTION
**Pull request by cynepco3hahue**
1. Added function to filesystem module
    - get_file_permissions
    - get_file_owner
    - is_user_exist
    - is_group_exist
2. Change function copy_to under host

### CODE REVIEW

File: rrmngmnt/filesystem.py [hunk 1]
```diff
@@ -161,3 +162,59 @@ def wget(self, url, output_file, progress_handler=None):
                 "Failed to download file from url {0}".format(url)
             )
         return output_file
+
+    def get_file_permissions(self, file_path):
```
Comment (lukas-bednar) [inline rrmngmnt/filesystem.py]: I would rather add method `stat` which behave same as
```
print os.stat("/tmp")
posix.stat_result(st_mode=17407, st_ino=15587, st_dev=38, st_nlink=26, st_uid=0, st_gid=0, st_size=920, st_atime=1462971342, st_mtime=1462976451, st_ctime=1462976451)
```

Comment (lukas-bednar) [inline rrmngmnt/filesystem.py]: something like:
```python
import subprocess
import collections

type_map = {
    'st_atime': ('%X', float),
    'st_blksize': ('%o', int),
    'st_blocks': ('%b', int),
    'st_ctime': ('%W', float),
    'st_dev': ('%d', int),
    'st_gid': ('%g', int),
    'st_ino': ('%i', int),
    'st_mode': ('0x%f', lambda x: int(x, 16)),
    'st_mtime': ('%Y', float),
    'st_nlink': ('%h', int),
    'st_rdev': ('%t', int),
    'st_size': ('%s', int),
    'st_uid': ('%u', int),
}

posix_stat_result = collections.namedtuple(
    'posix_stat_result', type_map.keys()
)

cmd = 'stat -c "%s" /tmp' % ",".join(
    [
        "%s=%s" % (k, v[0])
        for k, v in type_map.items()
    ]
)

p = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE)
out, _ = p.communicate()
out = out.strip().split(',')

data = {}

for pair in out:
    key, value = pair.split('=')
    data[key] = type_map[key][1](value)

result = posix_stat_result(**data)
print result
```
```
posix_stat_result(st_ctime=0.0, st_rdev=0, st_blocks=0, st_nlink=26, st_gid=0, st_dev=38, st_ino=15587, st_mode=17407, st_mtime=1462979364.0, st_blksize=4096, st_size=940, st_uid=0, st_atime=1462971342.0)
```

it looks like I have wrong c_time

**Comment by lukas-bednar** [comment]
please add tests

File: rrmngmnt/filesystem.py [hunk 2]
```diff
@@ -161,3 +162,59 @@ def wget(self, url, output_file, progress_handler=None):
...
+
+        :param file_path: absolute path to file
+        :type file_path: str
+        :return: file user and group owner names(example ['root', 'root'])
+        :rtype: list
+        """
+        cmd = ["stat", "-c", "\"%U %G\"", file_path]
+        return self._exec_command(cmd=cmd).split()
+
+    def is_user_exist(self, user_name):
```
Comment (lukas-bednar) [inline rrmngmnt/filesystem.py]: I am not sure whether these methods belongs to this module, it actually doesn't have nothing to do with filesystem

File: rrmngmnt/operatingsystem.py [hunk 3]
```diff
@@ -107,3 +110,97 @@ def distribution(self):
         if not self._dist:
             self._dist = self.get_distribution()
         return self._dist
+
+    def get_file_stats(self, path):
```
Comment (lukas-bednar) [inline rrmngmnt/operatingsystem.py]: just 'stat' as os.stat

File: rrmngmnt/operatingsystem.py [hunk 4]
```diff
@@ -107,3 +110,97 @@ def distribution(self):
...
+        return posix_stat_result(**data)
+
+    def get_file_permissions(self, path):
+        """
+        Get file permissions
+
+        :return: file permission in octal form(example 0644)
+        :rtype: str
+        """
+        return oct(self.get_file_stats(path=path).st_mode)[3:]
```
Comment (lukas-bednar) [inline rrmngmnt/operatingsystem.py]: looks like oct has diff behav in py2 & py3 ...

```
$ python -c "print oct(10)" 
012
$ python3 -c "print(oct(10))" 
0o12
```
please add correct handling

File: rrmngmnt/operatingsystem.py [hunk 5]
```diff
@@ -107,3 +110,97 @@ def distribution(self):
...
+        """
+        Get file user and group owner name
+
+        :return: file user and group owner names(example ['root', 'root'])
+        :rtype: list
+        """
+        cmd = ["stat", "-c", "%U %G", path]
+        return self._exec_command(cmd=cmd).split()
+
+    def is_user_exist(self, user_name):
```
Comment (lukas-bednar) [inline rrmngmnt/operatingsystem.py]: I would remove is_ prefix and add 's' at the end

File: tests/test_os.py [hunk 6]
```diff
@@ -175,3 +175,157 @@ def test_get_release_info(self):
...
+
+    def test_is_user_exist(self):
+        assert self.get_host().os.is_user_exist('root')
+
+    def test_is_group_exist(self):
+        assert self.get_host().os.is_group_exist('root')
+
+
+class TestFileStatsNegative(object):
+    type_map = {
```
Comment (lukas-bednar) [inline tests/test_os.py]: you can share that

File: rrmngmnt/host.py [hunk 7]
```diff
@@ -252,11 +253,26 @@ def copy_to(self, resource, src, dst):
...
+            path=src
+        )
         with resource.executor().session() as resource_session:
             with self.executor().session() as host_session:
                 with resource_session.open_file(src, 'rb') as resource_file:
                     with host_session.open_file(dst, 'wb') as host_file:
                         host_file.write(resource_file.read())
+        self.fs.chmod(path=dst, mode=file_permissions)
+        if (
+            self.os.is_user_exist(user_name=file_owner_user) or
```
Comment (lukas-bednar) [inline rrmngmnt/host.py]: I think the name of this method was changed

**Comment by lukas-bednar** [comment]
👍

File: rrmngmnt/host.py [hunk 8]
```diff
@@ -252,11 +253,26 @@ def copy_to(self, resource, src, dst):
...
                 with resource_session.open_file(src, 'rb') as resource_file:
                     with host_session.open_file(dst, 'wb') as host_file:
                         host_file.write(resource_file.read())
+        self.fs.chmod(path=dst, mode=file_permissions)
+        if (
+            self.os.user_exists(user_name=file_owner_user) or
+            self.os.group_exists(group_name=file_owner_group)
+        ):
+            if self.os.user_exists(user_name=file_owner_user):
+                self.fs.chown(path=dst, username=file_owner_user, groupname="")
```
Comment (lukas-bednar) [inline rrmngmnt/host.py]:
```
[root@lbednar tmp]# touch aaa
[root@lbednar tmp]# chown lbednar: aaa 
[root@lbednar tmp]# ls -la
-rw-r--r--.  1 lbednar lbednar     0 Jun  7 17:49 aaa
[root@lbednar tmp]# chown root: aaa
-rw-r--r--.  1 root    root        0 Jun  7 17:49 aaa
```
it changes both

Comment (lukas-bednar) [inline rrmngmnt/host.py]: I was thinking about it little bit and I think, it is not good idea to try set same ownership of file as it was on source machine.

so I suggest to do it like this:
```python
def copy_to(self, resource, src, dst, mode=None, ownership=None):
    # ...
    if mode:
        chmod(file, mode)
    if ownership:
        if not instance(ownership, tuple):
            raise TypeError("...")
        chown(file, *ownership)
    # ...
```
what do you think ?

Comment (cynepco3hahue) [inline rrmngmnt/host.py]: hm I believe it's ok, will leave choices to one who use function

File: rrmngmnt/host.py [hunk 9]
```diff
@@ -251,12 +252,20 @@ def copy_to(self, resource, src, dst):
...
         """
         with resource.executor().session() as resource_session:
             with self.executor().session() as host_session:
                 with resource_session.open_file(src, 'rb') as resource_file:
                     with host_session.open_file(dst, 'wb') as host_file:
                         host_file.write(resource_file.read())
+        if mode:
+            self.fs.chmod(path=dst, mode=mode)
+        if ownership:
+            self.fs.chown(path=dst, *ownership)
```
Comment (lukas-bednar) [inline rrmngmnt/host.py]: I think you can not pass keyword and than *args  ... I think it will end with TypeError

Comment (lukas-bednar) [inline rrmngmnt/host.py]: I just verified that
```
In [1]: def a(b, c, d):
   ...:     pass
   ...: 

In [2]: u = (2, 3)

In [3]: a(b=1, *u)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-4-53262b85573f> in <module>()
----> 1 a(b=1, *u)

TypeError: a() got multiple values for keyword argument 'b'
```

**Pull request merged**